transaction-pool/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use solana_sdk::pubkey::Pubkey;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+#[derive(Default, Clone, PartialEq, Debug)]
+pub struct Item {
+    pub priority: u64,
+    pub units: u64,
+    pub time: u64,
+    pub id: (u32, u32),
+}
+
+#[derive(Default, Clone)]
+pub struct Pool {
+    pub max_age: u64,
+    pub max_bucket_cu: u64,
+    pub max_block_cu: u64,
+    pub count: usize,
+    items: BTreeMap<u64, VecDeque<Item>>,
+}
+
+pub trait Table {
+    fn keys(&self, item: &Item) -> &[&Pubkey];
+}
+
+impl Pool {
+    pub fn insert(&mut self, item: Item) {
+        let bucket = self
+            .items
+            .entry(item.priority)
+            .or_insert_with(VecDeque::new);
+        bucket.push_back(item);
```

Comment:
So items are in time-order if same priority? In normal condition, assume most packets will fall into few buckets, to the extreme where a pool is just one VecDeque?

Reply:
yea, basically, and retry puts it in the back of the fifo queue.  not sure if that is right, or should be up front.  up front would preserve the order.